torch/fx/experimental/symbolic_shapes.py

```diff
@@ -3316,7 +3344,13 @@ def replace(expr, repl):
 
     @_lru_cache
     def replace(self, expr: "sympy.Expr") -> "sympy.Expr":
-        replacements = {s: self._find(cast(sympy.Symbol, s)) for s in expr.free_symbols}
+        replacements = {}
+        for s in expr.free_symbols:
+            new_s = self._find(cast(sympy.Symbol, s))
+            if new_s != s:
+                replacements[s] = new_s
+        if free_unbacked_symbols(expr):
+            replacements.update(self.complex_replacements)
```

Comment:
We should also do replacements of `self.replacements` in `self.complex_replacements`. We can do that while adding a new `complex_replacement`. Or is it an invariant that these replacements have already been applied? If so, can you assert that here?